### DISCUSSION TITLE
Does docmost "track" file attachments?

### DISCUSSION THREAD

**Discussion by ovizii** [docmost/docmost#39] [Help / Questions] [answered]
Say I upload attachments to a page, later rename and then delete the page, will the attachment persist? 
Can I somehow reuse an attachment uploaded to page A also on page B?

You get the gist, I'm interested how files or inline files are handled.

**Comment by Philipinho** [comment, marked as answer]
Attachments uploaded to a page will belong to the page. It will inherit the permissions of the page.

When you permanently delete a page, its attachments will be deleted alongside, but this is not yet enforced. 
We keep track of files attached to pages in the `attachments` table.

I would say, you should not reuse attachments on other pages. This is due to permissions enforcement reasons.

**Reply by ovizii** [reply]
Thanks. I was concerned because a competing product I tried before, did not "track" attachments. So when by chance you removed an inline image while editing, the only way to recover it was to re-upload. And that caused you to have the same attachment on your system twice as the removed one wasn't automatically deleted. :-/

**Reply by ovizii** [reply]
> When you permanently delete a page, its attachments will be deleted alongside, but this is not yet enforced.

Does that mean if I upload an attachment, then delete the page, I'm stuck with a non-attached file in the system I don't see and can't manage from within docmost?

**Reply by Philipinho** [reply]
1. I see. That's not the case with Docmost. 
2. The image is still tracked. You can run a script to fetch and delete attachments to pages that no longer exist.

Also, I plan to introduce a UI to view all attachments in a Space, which will include details of the uploader and the page it was attached to.  We already have the records in the database, it wouldn't hurt to make it visible to the right people.

**Reply by neo-jgrec** [reply]
Hey @Philipinho, what do you mean by fetching and deleting the attachments ? Do you have the script you're talking about with you ? That would be welcomed !

**Reply by Philipinho** [reply]
@neo-jgrec you can write a script to query the `attachments` table for attachments where the pages do not exist in the `pages` table.
```sql
SELECT a.id, a.file_name, a.file_path, a.space_id AS attachment_id, a.page_id
    FROM attachments a
    LEFT JOIN pages p ON a.page_id = p.id
    WHERE a.page_id IS NOT NULL
      AND p.id IS NULL
```
The SQL query above will list orphaned attachments.

You can then delete them from your storage driver (local or S3) based on the `file_path`.

**Reply by neo-jgrec** [reply]
Thank you, works like a charm!
